serializer_impl.hh

```diff
@@ -791,15 +791,56 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
 // using a range.
 // Use begin() and end() to iterate through the frozen vector,
 // deserializing (or skipping) one element at a time.
-template <typename T, typename InputStream = utils::input_stream>
+template <typename T, bool IsForward=true>
 class vector_deserializer {
 public:
     using value_type = T;
-    using input_stream = InputStream;
+    using input_stream = utils::input_stream;
 
 private:
     input_stream _in;
     size_t _size;
+    utils::chunked_vector<input_stream> _substreams;
+
+    size_t get_next_position(input_stream& in) {
+        serializer<T>::skip(in);
+        return in.position();
+    }
+    void fill_substreams() requires (!IsForward) {
+        input_stream in = _in;
+        input_stream in2 = _in;
+        size_t prev_position = in.position();
+        for (size_t i = 0; i < size(); ++i) {
+            size_t next_position = get_next_position(in);
+            size_t size = next_position - prev_position;
+            prev_position = next_position;
+
+            _substreams.push_back(in2.read_substream(size));
+        }
+    }
+
+    struct forward_iterator_data {
+        input_stream _in = simple_input_stream();
+        void skip() {
+            serializer<T>::skip(_in);
+            // auto len = read_frame_size();
+            // _in.skip(len);
+        }
+        value_type deserialize_next() {
+            return deserialize(_in, boost::type<T>());
+        }
+    };
+    struct reverse_iterator_data {
+        std::reverse_iterator<utils::chunked_vector<input_stream>::const_iterator> _substreams;
```

Comment:
The plural name is confusing... _substream in singular makes more sense since it is an iterator to the current substream

Reply:
Correct, forgot to replace it with something more sensible after changes.